src/inmanta/server/services/policy_engine_service.py

```diff
@@ -0,0 +1,199 @@
+from tornado.httpclient import HTTPRequest
+from tornado.simple_httpclient import SimpleAsyncHTTPClient
+
+from inmanta import config, const, server, util
+from inmanta.protocol.auth import decorators
+from inmanta.server import config as server_config
+from inmanta.server import protocol
+
+LOGGER = logging.getLogger(__name__)
+
+policy_file = config.Option(
+    "policy-engine", "policy-file", "/etc/inmanta/authorization/policy.rego", "File defining the access policy.", config.is_str
+)
+policy_engine_bind_address = config.Option(
+    "policy-engine",
+    "bind-address",
+    "127.0.0.1",
+    "Address on which the policy engine will listen for incoming connections.",
+    config.is_str,
+)
+policy_engine_bind_port = config.Option(
+    "policy-engine", "bind-port", 8181, "Port on which the policy engine will listen for incoming connections.", config.is_int
+)
+
+
+class PolicyEngineSlice(protocol.ServerSlice):
+
+    def __init__(self) -> None:
+        super().__init__(server.SLICE_POLICY_ENGINE)
+        self._opa_process = OpaServer()
+
+    async def start(self) -> None:
+        await super().start()
+        if server_config.enforce_access_policy.get():
+            LOGGER.info("Starting policy engine")
+            await self._opa_process.start()
+
+    async def stop(self) -> None:
+        await super().stop()
+        await self._opa_process.stop()
+
+    def get_depended_by(self) -> list[str]:
+        return [server.SLICE_TRANSPORT]
+
+    async def does_satisfy_access_policy(self, input_data: Mapping[str, object]) -> bool:
+        """
+        Return True iff the policy evaluates to True.
+        """
+        if not server_config.enforce_access_policy.get():
+            return True
+        if not self._opa_process.running:
+            raise Exception("Policy engine is not running. Call OpaServer.start() first.")
+        client = SimpleAsyncHTTPClient()
+        policy_engine_addr = await self._opa_process.get_addr_policy_engine()
+        request = HTTPRequest(
+            url=f"http://{policy_engine_addr}/v1/data/policy/allowed",
+            method="POST",
+            headers={"Content-Type": "application/json"},
+            body=json.dumps(input_data, default=util.api_boundary_json_encoder),
+        )
+        try:
+            response = await client.fetch(request)
+            if response.code != 200:
+                LOGGER.error("Failed to evaluate access policy for %s.", input_data)
+                return False
+            response_body = json.loads(response.body.decode())
+            return "result" in response_body and response_body["result"] is True
+        except Exception:
+            LOGGER.exception("Failed to evaluate access policy for %s.", input_data)
+            return False
+
+
+class OpaServer:
+
+    def __init__(self) -> None:
+        self.process: asyncio.subprocess.Process | None = None
+        self.running = False
+
+    async def get_addr_policy_engine(self) -> str:
```

Comment:
why is this async?

Reply:
No specific reason. Do you know what the general development guideline is for this case?